workers/social/lib/social/models/gitprovider/gitlabprovider.coffee

```diff
@@ -1,72 +1,167 @@
 Constants   = require './constants'
 helpers     = require './utils/helpers'
 async       = require 'async'
+request     = require 'request'
 URL         = require 'url'
 _           = require 'lodash'
-GitlabAPI   = require 'node-gitlab'
+GitlabAPI   = require 'gitlab'
 KodingError = require '../../error'
+KONFIG      = require 'koding-config-manager'
+
+
+getPrivateToken = (options, callback) ->
+
+  { urlData: { baseUrl }, token, juser } = options
+
+  params    =
+    url     : "#{baseUrl}/api/v3/user"
+    headers : { Authorization : "Bearer #{token}" }
+
+  request.get params, (err, res, body) ->
+    return callback err  if err
+
+    try
+      data  = JSON.parse body
+      token = data.private_token
+    catch e
+      return callback new KodingError 'Auth failed'
+
+    unless token
+      console.log '[GITLAB] TOKEN NOT FOUND ON DATA:', data
+      return callback new KodingError 'Auth failed'
+
+    juser.update {
+      $set: {
+        'foreignAuth.gitlab.privateToken': token
+      }
+    }, (err) ->
+
+      console.warn 'Error while updating private token:', err  if err
+
+      # Swallow update error here thus we don't need it ~GG
+      callback null, token
+
 
 module.exports = GitLabProvider =
 
+
   importStackTemplateData: (importParams, user, callback) ->
 
     { url, privateToken } = importParams
     return  unless urlData = @parseImportUrl url
 
-    if privateToken
+    if privateToken or privateToken = user.getAt 'foreignAuth.gitlab.privateToken'
       @importStackTemplateWithPrivateToken privateToken, urlData, callback
+    else if oauth = user.getAt 'foreignAuth.gitlab'
+      @importStackTemplateWithOauth oauth, urlData, user, callback
     else
       @importStackTemplateWithRawUrl urlData, callback
 
     return yes
 
 
+  parseRepo: (url) ->
+
+    { gitlab } = KONFIG
+    [user, repo, branch] = url.split '/'
+
+    port    = if port = gitlab.port then ":#{port}" else ''
+    baseUrl = "http://#{gitlab.host}#{port}" # FIXME update protocol here ~GG
+    branch ?= 'master'
+    url     = "#{baseUrl}/#{user}/#{repo}"
+
+    if branch isnt 'master'
+      url = "#{url}/tree/#{branch}"
+
+    return { originalUrl : url, baseUrl, user, repo, branch }
+
+
   parseImportUrl: (url) ->
 
+    # if user/repo/branch provided as url we will use
+    if not (/^http(|s)\:\/\//.test url) and (url.split '/').length in [2, 3]
```

Comment:
no worries :) I'll get all details as query strings instead, in this way it's not extendable.